service/history/historyEngine.go

```diff
@@ -2357,9 +2383,14 @@ func (e *historyEngineImpl) TerminateWorkflowExecution(
 		domainID,
 		workflowExecution,
 		e.timeSource.Now(),
-		func(wfContext execution.Context, mutableState execution.MutableState) (*workflow.UpdateAction, error) {
+		func(ctx context.Context, wfContext execution.Context, mutableState execution.MutableState) (*workflow.UpdateAction, error) {
 			if !mutableState.IsWorkflowExecutionRunning() {
-				return nil, workflow.ErrAlreadyCompleted
+				return nil, execution.VersionBasedError(
+					workflow.ErrAlreadyCompleted,
+					"1.7.0",
```

Comment:
let's define this value in `versionChecker.go`.

Reply:
I am still thinking maybe we should do this check in frontend instead of history so that only frontend interacts with the concept of client version. Then we can add a new method to versionChecker for checking the support of already complete error at client side. 

In terms of implementation, one approach might be, history always return the newly defined type (just in 3 APIs for now), and frontend can convert it back to entity not exists error if necessary.

Reply:
That sounds like a good idea. Do we know which client is calling the service? How do I know if the client is java or go client?